sql/src/main/java/io/crate/operation/scalar/ArrayDifferenceFunction.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.scalar;
+
+import io.crate.metadata.*;
+import io.crate.operation.Input;
+import io.crate.planner.symbol.Function;
+import io.crate.planner.symbol.Literal;
+import io.crate.planner.symbol.Symbol;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.elasticsearch.common.Preconditions;
+
+import java.util.*;
+
+public class ArrayDifferenceFunction extends Scalar<Object[], Object> {
+
+    public static final String NAME = "array_difference";
+    private FunctionInfo functionInfo;
+    private Set<Object> subtractSet;
+
+    private static FunctionInfo createInfo(List<DataType> types) {
+        ArrayType arrayType = (ArrayType) types.get(0);
+        if(arrayType.innerType().equals(DataTypes.UNDEFINED)){
+            arrayType = (ArrayType) types.get(1);
+        }
+        return new FunctionInfo(new FunctionIdent(NAME, types), arrayType);
+    }
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(NAME, new Resolver());
+    }
+
+    protected ArrayDifferenceFunction(FunctionInfo functionInfo) {
+        this.functionInfo = functionInfo;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return functionInfo;
+    }
+
+    @Override
+    public Symbol normalizeSymbol(Function symbol) {
+        if (anyNonLiterals(symbol.arguments())) {
+            return symbol;
+        }
+
+        Input[] argList = new Input[symbol.arguments().size()];
+        int counter = 0;
+        for(Symbol argSymbol : symbol.arguments()){
+            Input input  = (Input) argSymbol;
+            argList[counter++] = input;
+        }
+
+        return Literal.newLiteral(evaluate(argList), this.info().returnType());
+    }
+
+    @Override
+    public Scalar<Object[], Object> compile(List<Symbol> arguments) {
+
+        subtractSet = new HashSet<>();
```

Comment:
please only initialize ``subtractSet`` when successfully compiled.
Even better: use ``private final Optional<Set<?>> subtractSet`` to hold the subtractSet if available and return a new instance with the Set here in ``.compile()`` if compilation works. If not, simply return this. avoid mutable state if possible. As a kind of inspiration: https://github.com/crate/crate/pull/2629/files#diff-4231faae44236f1fede93f2494d98b2a